spyder/api/plugins.py

```diff
@@ -533,3 +557,874 @@ def set_ancestor(self, ancestor):
+    # Example: ['plots', 'ipython_console', ...]
+    REQUIRES = None
+
+    # List of required plugin dependencies names
+    # Example: ['plots', 'ipython_console', ...]
+    OPTIONAL = None
+
+    # This is the a widget (or any non widget class) to store behavior and
+    # actions. Subclass SpyderWigdetMixin
+    WIDGET_CLASS = None
+
+    # Name of the configuration section that's going to be
+    # used to record the plugin's permanent data in Spyder
+    # config system (i.e. in spyder.ini)
+    CONF_SECTION = None
+
+    # Use separate configuration file for plugin
+    CONF_FILE = True
+
+    # Define configuration defaults if using a separate file.
+    # List of tuples, with the first item in the tuple being the section
+    # name and the second item being the default options dictionary.
+    #
+    # CONF_DEFAULTS_EXAMPLE = [
+    #     ('section-name', {'option-1': 'some-value',
+    #                       'option-2': True,}),
+    #     ('another-section-name', {'option-3': 'some-other-value',
+    #                               'option-4': [1, 2, 3],}),
+    # ]
+    CONF_DEFAULTS = None
+
+    # Define configuration version if using a separate file
+    #
+    # IMPORTANT NOTES:
+    # 1. If you want to *change* the default value of a current option, you
+    #    need to do a MINOR update in config version, e.g. from 3.0.0 to 3.1.0
+    # 2. If you want to *remove* options that are no longer needed or if you
+    #    want to *rename* options, then you need to do a MAJOR update in
+    #    version, e.g. from 3.0.0 to 4.0.0
+    # 3. You don't need to touch this value if you're just adding a new option
+    CONF_VERSION = None
+
+    # Widget to be used as entry in Spyder Preferences dialog
+    CONF_WIDGET_CLASS = None
+
+    # Path for images relative to the plugin path
+    IMG_PATH = 'images'
+
+    # Control the font size relative to the global fonts defined in Spyder
+    FONT_SIZE_DELTA = 0
+    RICH_FONT_SIZE_DELTA = 0
+
+    # --- API: Signals -------------------------------------------------------
+    # ------------------------------------------------------------------------
+
+    # Use this signal to inform another plugin that configuration changed.
+    sig_free_memory_requested = Signal()
+    sig_option_changed = Signal(str, object)
+    sig_quit_requested = Signal()
+    sig_redirect_stdio_requested = Signal(bool)
+    sig_restart_requested = Signal()
+    sig_shortcut_registration_requested = Signal(QObject, str, str, bool)
+    sig_status_message_requested = Signal(str, int)
+
+    # --- Private attributes -------------------------------------------------
+    # ------------------------------------------------------------------------
+    # Define configuration name map for plugin to split configuration
+    # among several files. See spyder/config/main.py
+    _CONF_NAME_MAP = None
+
+    # Store reference to registered plugins in a dictionary
+    # {plugin.NAME: plugin_instance, ...}
+    _PLUGINS = None
```

Comment:
Shouldn't this be an attribute of the main window? It seems you're using it to access the plugins registered before, but I don't understand how you're doing that if this an attribute of the class.

Reply:
Yes at some point it will need to be a part of the `PluginManager` which will live in the MainWindow. But that will only be posible after ALL plugins are migrated